src/v/cluster/tests/id_allocator_stm_test.cc

```diff
@@ -22,112 +23,161 @@
 #include "test_utils/async.h"
 #include "test_utils/fixture.h"
 #include "test_utils/scoped_config.h"
+#include "tests/stm_test_fixture.h"
 
 #include <seastar/core/abort_source.hh>
 #include <seastar/core/future.hh>
 #include <seastar/core/sleep.hh>
 #include <seastar/util/defer.hh>
 #include <seastar/util/log.hh>
 
-#include <boost/range/irange.hpp>
-#include <boost/test/tools/old/interface.hpp>
-
 #include <thread>
 
 using namespace std::chrono_literals;
 
+namespace cluster {
+namespace {
+
 ss::logger idstmlog{"idstm-test"};
 
-struct id_allocator_stm_fixture : simple_raft_fixture {
-    void create_stm_and_start_raft() {
+struct id_allocator_stm_fixture : state_machine_fixture {
+    ss::future<> initialize_state_machines() {
         // set configuration parameters
         test_local_cfg.get("id_allocator_batch_size").set_value(int16_t(1));
         test_local_cfg.get("id_allocator_log_capacity").set_value(int16_t(2));
-        create_raft();
-        raft::state_machine_manager_builder stm_m_builder;
 
-        _stm = stm_m_builder.create_stm<cluster::id_allocator_stm>(
-          idstmlog, _raft.get(), config::shard_local_cfg());
+        create_nodes();
+        co_await start_nodes();
+    }
+
+    ss::future<> start_nodes() {
+        co_await parallel_for_each_node(
+            [this](raft_node_instance& node) -> ss::future<> {
+                co_await node.initialise(all_vnodes());
+                raft::state_machine_manager_builder builder;
+                auto stm = builder.create_stm<cluster::id_allocator_stm>(idstmlog, node.raft().get(), config::shard_local_cfg());
+                co_await node.start(std::move(builder));
+                node_stms.emplace(node.get_vnode(), std::move(stm));
+            }
+        );
+    }
+
+    ss::future<> stop_and_recreate_nodes() {
+        absl::flat_hash_map<model::node_id, ss::sstring> data_directories;
+        for (auto& [id, node] : nodes()) {
+            data_directories[id] = node->raft()->log()->config().base_directory();
+            node_stms.erase(node->get_vnode());
```

Comment:
This one worries me. It would not make sense to emplace them again without erasing, as emplace does not replace existing entries. Keeping the old ones results in errors. But another test [somehow gets away with this](https://github.com/redpanda-data/redpanda/blob/dev/src/v/raft/tests/persisted_stm_test.cc#L401-L412), which questions my understanding of the whole thing.

Reply:
We discussed this. This is a bug in `persisted_stm` test. The test doesn't fail as the old stm instance is up to date with the state and this is what is being tested after the restart. The test validates if the state machine can recover the state after restart.